lib/puppet/x509/cert_provider.rb

```diff
@@ -146,14 +147,33 @@ def load_private_key(name, required: false)
   # Load a PEM encoded private key.
   #
   # @param pem [String] PEM encoded private key
-  # @return [OpenSSL::PKey::RSA] The private key
-  # @raise [OpenSSL::PKey::RSAError] The `pem` text does not contain a valid key
+  # @return [OpenSSL::PKey::RSA, OpenSSL::PKey::EC] The private key
+  # @raise [OpenSSL::PKey::PKeyError] The `pem` text does not contain a valid key
   # @api private
   def load_private_key_from_pem(pem)
     # set a non-nil passphrase to ensure openssl doesn't prompt
     # but ruby 2.4.0 & 2.4.1 require at least 4 bytes, see
     # https://github.com/ruby/ruby/commit/f012932218fd609f75f9268812df61fb26e2d0f1#diff-40e4270ec386990ac60d7ab5ff8045a4
-    OpenSSL::PKey::RSA.new(pem, '    ')
+    if Puppet::Util::Platform.jruby?
+      begin
+        if pem =~ EC_HEADER
+          OpenSSL::PKey::EC.new(pem, '    ')
+        else
+          OpenSSL::PKey::RSA.new(pem, '    ')
+        end
+      rescue OpenSSL::PKey::PKeyError => e
+        if e.message =~ /Neither PUB key nor PRIV key/
+          raise OpenSSL::PKey::PKeyError, "Could not parse PKey: no start line"
+        else
+          raise e
+        end
+      end
+    else
+      OpenSSL::PKey.read(pem, '    ')
+    end
+  rescue ArgumentError => e
+    # handle EC keys on ruby <= 2.3
```

Comment:
This comment seems odd - is this rescue expected with EC keys on ruby <= 2.3?  Are we not expecting to support EC keys for those earlier ruby versions?  Or maybe I'm not understanding the comment.

Reply:
Ah, it's because ruby <= 2.3 raises an `ArgumentError` if given an encrypted EC or RSA key instead of instead of the more standard `OpenSSL::PKey::PKeyError`. For example, if I remove that line then 2 tests fail with:

```
 expect {
         provider.load_private_key('encrypted-key')
       }.to raise_error(OpenSSL::PKey::PKeyError, /Could not parse PKey: no start line/)

       expected OpenSSL::PKey::PKeyError with message matching /Could not parse PKey: no start line/, got #<ArgumentError: Could not parse PKey: no start line> with backtrace:
```
Let me update the comment, because it's not specific to EC.

Reply:
This issue was fixed in ruby 2.4 (https://bugs.ruby-lang.org/issues/11774). I monkey patched `OpenSSL::PKey.read` when running on ruby 2.3 (since we don't support earlier ruby versions).